File: searching_algo/index_sequential_search.c
// Index Sequential Search
// 색인 순차 탐색

#include <stdio.h>

#define MAX_SIZE 1000
#define INDEX_SIZE 10

int n = 9;
int list[MAX_SIZE] = {3, 9, 15, 22, 31, 55, 67, 88, 91};

// making index table structure
typedef struct {
	int key;
	int index;
} indexTable;

// index table
indexTable index_list[INDEX_SIZE] = { {3, 0}, {15, 2}, {67, 6} };

// Search in list between two index : LOW & HIGH
int sequential_search(int key, int low, int high) {
	int i;
	for (i = low; i <= high; i++) {
		if (list[i] == key) {
			printf("Key is in index No.%d.\n", i);
			return i;
		}
	}
	return -1;
}

// Search index in INDEX TABLE
int index_search(int key) {
	int i, low, high;

	// if key is not in list, searching gone break
	if (key < list[0] || key > list[n - 1]) {
		return -1;
	}
	
	// search index table to find range of key
	for (i = 0; i < INDEX_SIZE; i++) {
		if (index_list[i].key <= key && index_list[i + 1].key > key) {
			break;
		}
	}
	
	// result of "i" after above loop
	if (i == INDEX_SIZE) {
		low = index_list[i - 1].index;
		high = n;
		printf("Key is located in between of No.%d INDEX and No.%d INDEX.\n", low, high);

	} else {
		low = index_list[i].index;
		high = index_list[i + 1].index;
		printf("Key is located in between of No.%d INDEX and No.%d INDEX.\n", low, high);
	}

	return sequential_search(key, low, high);
}

int main() {
	index_search(31);
}

File: searching_algo/index_sequential_search.py
# Index Sequential Search
# 색인 순차 탐색

index_size = 3

num_list = [3, 9, 15, 22, 31, 55, 67, 88, 91]
index_table = [[0, 3], [2, 15], [6, 67]]

def sequential_search(key, low, high):
    for i in range(low, high, 1):
        if num_list[i] == key:
            print("Index number :", i, "Key :", key)
            return i

    return -1

def index_search(key):

    i, low, high = 0, 0, 0

    if key < num_list[0] or key > num_list[len(num_list) - 1]:
        return -1
    
    for i in range(0, index_size - 1, 1):
        if index_table[i][0] <= key and index_table[i + 1][0] > key:
            break

    if i == index_size:
        low = index_table[i - 1][0]
        high = index_size - 1

    else:
        low = index_table[i][0]
        high = index_table[i + 1][0]

    return sequential_search(key, low, high)


index_search(88)
